### PR TITLE
feat(build)!: use tsup

### LINKED ISSUE #50: 💡 Improvement: use esbuild via tsup (or swc)
[`tsup`](https://github.com/egoist/tsup) is a very powerful build tool that harvests esbuild to compile typescript _very_ quickly (like, subsecond crazy). 

**Current Issue**
`tsc` is really slow. It's written in TS, which means it doesn't have the power as a go or rust backed build tool. There's even [plans to outright rewrite it](https://kdy1.dev/posts/2022/1/tsc-go) in Go. Buuut until that happens, nothing we can do as we're growing more and more. 

**The Solution**
Replace all `tsc` with `tsup`. We'll stay fast forever (so as long as we don't use decorators 😬) with the added benefit of tsup being able to do clean builds, restart on success through `--onSuccess`, and whatever else esbuild can do in the future. 

**The Drawbacks**
This is kinda silly right? Spending 10 minutes installing and configuring esbuild for what? 800ms improvement? I thought this at first as well. But after testing it on my bot (which sadly uses decorators 😬) - which is quite large, it was obvious that tsup improves developer experience not just with the improved build speeds, but the `onSuccess` helper and auto-clean builds. Those are 2 things that are done manually at the moment, see [the package.json](https://github.com/MarcusOtter/discord-needle/blob/main/package.json#L7) to get an idea. This could be replaced _by the compiler_. Good argument there, then it would work on all operating systems 🥳!!

**The Alternatives**
Like every developer in the world, apparently 200ms build times is still "too slow", so they invented `@swc/core` (Hey wait a second, this is the same guy who wanted to rewrite `tsc` in `Go`...). Wow. `swc` is written in Rust, and it's _super fast_. Stupidly fast for us. I compared them on my bot and found a **71 millisecond difference**. Wait, that's not a lot.. 🤔.... Right! They're both plenty fast for us. So the argument switches from "Which one is faster?!" to "Which one has the best DX" and that's by far `tsup`. Though, if we ever touch decorators, `@swc/core` has support for them and likely will have more support around it in general. 

**My take**
I think this is only a benefit. As Needle grows, there's going to be more and more TS to compile. With that, comes the annoying reality of waiting slightly longer each time. Making this switch now will eliminate that worry and provides us with super speedy builds, no matter how large this project grows. Take that with the improved developer experience and I think it's a sell. This is also relatively low cost, as most of the setup for esbuild is hidden away in a neat, type-safe package.

### PR DESCRIPTION
**Pull request by c43721**
Fixes #50 

Build goes from 2 seconds to 20ms. Also cleans `dist` each time you run build.

Breaking changes: 
uses esbuild to build instead of tsc
~~dev command **no longer registers commands**. It will auto-recompile + restart the bot if changes are made.~~

### CODE REVIEW

**Comment by MarcusOtter** [comment]
This looks super cool, I'll make sure to give tsup a proper read through and test this out before trying, but awesome work!

> dev command no longer registers commands

This is actually pretty important since the commands won't update and you can't try the bot otherwise (if the commands are being developed). How do you think we should solve this? 🤔 Hot reloading seems awesome, I just wish there was a way we could have the best of both worlds.

I guess one very hacky thing I have in mind is to read what the current commands are and store that in a json file, then every hot reload it will read that object and see if anything needs to be changed, and if so deploy the new changes to the development guild.

**Comment by c43721** [comment]
Keeping track of command data is super annoying, so I forwent hot reloading commands since it could lead to API absue (think- you change a typo on a command then started to work on logic for that command).

We'd have to do something like sapphire/framework does and auto register based on what API responds. That's doable but requires custom code, not something I wish to do.

**Comment by c43721** [comment]
alright technically I can copy/paste their code but would mean that we'd have 2 licensed since they use MIT

**Comment by MarcusOtter** [comment]
I don't think that matters @c43721 

From MIT:
> Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, **sublicense**, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

As long as we keep the copyright & permission notice in the file, we are allowed to sublicense the code.

**Comment by c43721** [comment]
shows how much I know about licenses

**Comment by c43721** [comment]
this now registers commands on startup, but does not hot reload. There's a `deploy:dev` command to deploy development commands from scripts (since dev registers once + watches build)

**Comment by c43721** [comment]
never make me merge changes again...
@MarcusOtter Is there a chance this could get a review this week? @nchristopher anything changes you would like by chance?

**Comment by nchristopher** [comment]
I get this error when trying to run Needle.
```
npm run dev

> discord-needle@1.0.0 dev
> npm run deploy:dev && tsup --watch --onSuccess "node --enable-source-maps .

sh: 1: Syntax error: Unterminated quoted string
```

File: package.json [hunk 1]
```diff
@@ -4,11 +4,12 @@
...
   "main": "./dist/index.js",
   "scripts": {
-    "build": "rd /s /q dist & tsc",
-    "start": "npm run build && node ./dist/index.js",
-    "dev": "npm run build && node ./scripts/deploy-commands.js && node ./dist/index.js",
-    "undeploy": "npm run build && node ./scripts/deploy-commands.js  --undeploy",
-    "deploy": "npm run undeploy && node ./scripts/deploy-commands.js --global"
+    "build": "tsup",
+    "start": "npm run build && node --enable-source-maps .",
+    "dev": "npm run deploy:dev && tsup --watch --onSuccess \"node --enable-source-maps .",
```
Comment (nchristopher) [inline package.json:9]: This string isn't terminated, resulting in an error when running `npm run dev`.
```suggestion
    "dev": "npm run deploy:dev && tsup --watch --onSuccess \"node --enable-source-maps .\"",
```

Comment (c43721) [reply package.json]: 🤦

**Comment by MarcusOtter** [comment]
> never make me merge changes again...

oops 😄 I can fix the conflicts and review this week hopefully, I was unsure about the status of this PR which is why I haven't touched it

**Comment by c43721** [comment]
Good for a review now, FYI.

**Review by nchristopher — CHANGES_REQUESTED** [review]
It seems that `config.json` is being built as `config.js`, which causes imports to fail.

**Comment by c43721** [comment]
it should work :/

I didn't test this so I'll take another look eventually

**Comment by c43721** [comment]
@MarcusOtter Has merge conflicts :(

**Comment by nchristopher** [comment]
`/scripts/deploy-commands.js` needs to be modified to only use `.js` files otherwise it will attempt to deploy commands from `.map`.

**Comment by MarcusOtter** [comment]
> `/scripts/deploy-commands.js` needs to be modified to only use `.js` files otherwise it will attempt to deploy commands from `.map`.

I did this already with #100

**Comment by c43721** [comment]
tsconfig cannot have baseUrl for this to work

**Comment by nchristopher** [comment]
@c43721 Your fork's branch needs to be updated in order to fix command deploys.

File: tsup.config.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+export default defineConfig({
+	clean: true,
+	bundle: false,
+	dts: false,
+	entry: ["src/**/*.ts", "src/config.json"],
+	format: ["cjs"],
+	minify: false,
+	tsconfig: "tsconfig.json",
+	target: "esnext",
```
Comment (nchristopher) [inline tsup.config.ts:11]: We should explicitly define this.
```suggestion
	target: "node16.14",
```

File: src/helpers/configHelpers.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@
 // ________________________________________________________________________________________________
 
 import type { Client, Guild, ThreadChannel } from "discord.js";
-import * as defaultConfig from "../config.json";
+import * as defaultConfig from "../../config.json";
```
Comment (nchristopher) [inline src/helpers/configHelpers.ts:19]: ❗This is a breaking change, the next release to include this PR must be a major (eg. `3.0.0`) one.

**Review by nchristopher — CHANGES_REQUESTED** [review]

File: src/helpers/configHelpers.ts (hunk 3 above)
Comment (c43721) [reply src/helpers/configHelpers.ts:19]: hope next release is v3 anyways 🙂

File: src/commands/configure.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@
 // ________________________________________________________________________________________________
 
 import { SlashCommandBuilder } from "@discordjs/builders";
-import { ChannelType } from "discord-api-types";
+import { ChannelType } from "discord-api-types/v9";
```
Comment (nchristopher) [inline src/commands/configure.ts:19]: This change should be done everywhere else we import `discord-api-types`.

**Review by nchristopher — CHANGES_REQUESTED** [review]

File: src/commands/configure.ts (hunk 4 above)
Comment (c43721) [reply src/commands/configure.ts:19]: We should. I would like to do this in a follow up though

Comment (nchristopher) [reply src/commands/configure.ts:19]: This should be done in #94.

**Comment by MarcusOtter** [comment]
I'm not sure what the state of this is, should we turn it into a draft for now? Does it need my review? Last thing I heard was something about fixing imports, otherwise builds break, does that still apply?

**Comment by c43721** [comment]
Something something JSON.. I'll close since this is so far behind and I'll re-open/

**Pull request closed without merging**